src/github_runner_image_builder/utils.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Utilities used by the app."""
+
+import functools
+import logging
+import time
+from typing import Callable, Optional, Type, TypeVar
+
+from typing_extensions import ParamSpec
+
+logger = logging.getLogger(__name__)
+
+
+# Parameters of the function decorated with retry
+ParamT = ParamSpec("ParamT")
+# Return type of the function decorated with retry
+ReturnT = TypeVar("ReturnT")
+
+
+# This decorator has default arguments, one extra argument is not a problem.
+def retry(  # pylint: disable=too-many-arguments
```

Comment:
There are many retry pip libraries, couldn't we use one instead of defining our own decorator ? (for example https://github.com/jd/tenacity)

Reply:
I can create a followup PR to address retry and click :)

Reply:
Ok for retry, see my comment for click :)

Reply:
I do want to follow your and David's advice on having the application do uploading as well when the testing becomes a little easier so I think it's going to be worth it! Thank you for the suggestion 👍 ⚡